lektor/db.py

```diff
@@ -1283,6 +1288,22 @@ def _iter_datamodel_choices(datamodel_name, path, is_attachment=False):
     yield "none"
 
 
+def get_default_slug(record, type_=None):
+    """Compute the default slug for a page.
+
+    This computes the default value of ``_slug`` for a page.  The slug
+    is computed by expanding the parent’s ``slug_format`` value.
+
+    """
+    parent = getattr(record, "parent", None)
+    if parent is None:
+        return ""
+    return parent.datamodel.get_default_child_slug(record.pad, record)
+
+
+default_slug_descriptor = property(get_default_slug)
```

Comment:
Is there a specific reason why you moved this out of the class? In particular having it defined as a property outside of the class seems odd.

Reply:
Background: What we're trying to do here is defer the computation of the default slug until later.  To do that we put a descriptor — an object which as a callable `.__get__` attribute — the the records `data["slug"]`.  When the slug is first accessed via [Record.__getitem__](https://github.com/lektor/lektor/blob/46b28df9f97e4b03382f7d33b8323f7343c58672/lektor/db.py#L461) the `.__get__` will be called to compute the actual value of the data field.
This "deferred data field" API just happens to look very much like Python’s [attribute descriptor API](https://docs.python.org/3/howto/descriptor.html).

I had originally coded this section something like:

```python
class DefaultSlugDescriptor(object):
    def __get__(self, record, type_=None):
        parent = getattr(record, "parent", None)
￼        if parent is None:
￼            return ""
￼        return parent.datamodel.get_default_child_slug(record.pad, record)

default_slug_descriptor = DefaultSlugDescriptor()
```

I refactored this because I thought using `property` to generate the descriptor was more succinct and a bit clearer, but the two approaches are equivalent.

[As an aside, in the above example the unused `type_` argument is included to exactly match Python’s attribute descriptor API.  I don't think lektor ever invokes `__get__` with the second argument, so it could be omitted, but I copied the style of, e.g., [MarkdownDescriptor](https://github.com/lektor/lektor/blob/46b28df9f97e4b03382f7d33b8323f7343c58672/lektor/types/formats.py#L9) here.
In reference to your previous comment, I erroneously copied the phantom `type_` argument when I refactored to using `property` to construct the descriptor.]

Now to get to your question...

We are not using `property` to create a property of class `Database`, so it's use would seem odd whether left in the class or not.  I.e. we could do something like:

```python
class Database(object):
[...]
    def get_default_slug(self, record):
        # Note: method signature changed.  Used to be get_default_slug(self, data, pad)
        [...]

    def process_data(self, data, datamodel, pad):
￼        if not data["_slug"]:
￼            data["_slug"] = property(self.get_default_slug)
￼        else:
￼            [...]
```
(Note also that the `get_default_slug`, though not explicitly marked so, has always been a static method — it doesn't depend on or mutate the state of the `Database` instance.)

Anyhow, I thought it was clearer and cleaner outside of the class.  If you like another alternative better, I'd be happy to change it.

Reply:
Thanks for the explanation :)

> Anyhow, I thought it was clearer and cleaner outside of the class.

I agree in principle. However, I think in this instance, since moving it out of the class moves it a few hundred lines up in this rather monstrous file, it might actually be easier to keep it in the class (as a `@staticmethod` starting with an underscore). But if you'd rather have it outside the class that's fine too (maybe add an underscore), I can't really make up my mind whether I strongly prefer either.